src/Raven.Server/Documents/Replication/DocumentReplicationExecuter.cs

```diff
@@ -16,69 +18,110 @@
+                    else
+                        maxReceivedChangeVector.Add(currentEntry.DbId,currentEntry.Etag);
+                }
+
                 ReceiveReplicated(context, doc);
             }
 
+            for (int i = 0; i < dbChangeVector.Length; i++)
+            {
+                var receivedVal = maxReceivedChangeVector.GetOrAdd(dbChangeVector[i].DbId, 0);
+                if (receivedVal > dbChangeVector[i].Etag)
+                {
+                    changeVectorUpdated = true;
+                    dbChangeVector[i].Etag = receivedVal;
+                }
+            }
+
             if(changeVectorUpdated)
                 _database.DocumentsStorage.SetChangeVector(context,dbChangeVector);
         }
 
-        private bool UpdateDbChangeVectorIfNeededFrom(BlittableJsonReaderObject doc,ChangeVectorEntry[] dbChangeVector)
+        public override int GetHashCode()
         {
-            var docChangeVector = doc.EnumerateChangeVector();
-            var wasUpdated = false;
-            foreach (var entry in docChangeVector)
-            {
-                var indexOfDbEntry = dbChangeVector.IndexOf(e => e.DbId == entry.DbId);
-                if (indexOfDbEntry != -1 && dbChangeVector[indexOfDbEntry].Etag < entry.Etag)
-                {
-                    dbChangeVector[indexOfDbEntry].Etag = entry.Etag;
-                    wasUpdated = true;
-                }
-            }
+            return _srcDbId.GetHashCode() ^ _destination.Url.GetHashCode();
+        }
 
-            return wasUpdated;
+        public override bool Equals(object obj)
+        {
+            var otherExecuter = obj as DocumentReplicationExecuter;
+            if (otherExecuter == null)
+                return false;
+            return _srcDbId == otherExecuter._srcDbId &&
+                _destination.Url.Equals(otherExecuter._destination.Url, StringComparison.OrdinalIgnoreCase);
         }
 
         protected override void ExecuteReplicationOnce()
         {
-            
+            var lastSendEtag = _lastEtag;
+
+            //just for shorter code
+            var documentStorage = _database.DocumentsStorage;
+            //TODO: handle here properly last etag
+            //either add here negotiation for the etag, 
+            //or add etag tracking
+            DocumentsOperationContext context;			
+            using (documentStorage.ContextPool.AllocateOperationContext(out context))
+            using (context.OpenReadTransaction())
+            {
+                //TODO: make replication batch size configurable
+                //also, perhaps there should be timers/heuristics
+                //that would dynamically resize batch size
+                var replicationBatch =
+                    documentStorage
+                        .GetDocumentsAfter(context, lastSendEtag, 0, 1024)
+                        .ToArray();
+                AsyncHelpers.RunSync(() => _transport.SendDocumentBatchAsync(replicationBatch, context));
```

Comment:
Do not run code link this. We should reserver `RunSync` purely for the client side, where we have the sync API and just want to translate it from the async.
We should not be using this in the server at all